### PR TITLE
feat(core-flows,types,promotion): register promotion campaign usage upon cart completion

### PR DESCRIPTION
**Pull request by riqwan**
what:

- registers promotion campaign usage upon cart completion

RESOLVES CC-419

### CODE REVIEW

File: integration-tests/modules/__tests__/cart/store/carts.spec.ts [hunk 1]
```diff
@@ -2015,16 +2053,16 @@ medusaIntegrationTestRunner({
             type: "order",
             order: expect.objectContaining({
               id: expect.any(String),
-              total: 106,
+              total: 94.764,
```
Comment (riqwan) [inline integration-tests/modules/__tests__/cart/store/carts.spec.ts:2056]: @carlos-r-l-rodrigues can you help check the totals for a cart after applying a promo? does this look correct?

File: packages/core/types/src/promotion/service.ts [hunk 2]
```diff
@@ -32,26 +33,42 @@ export interface IPromotionModuleService extends IModuleService {
...
    *     amount: 5000,
    *     code: "FREESHIPPING",
    *   },
    * ])
    */
-  registerUsage(computedActions: ComputeActions[]): Promise<void>
+  registerUsage(computedActions: UsageComputedActions[]): Promise<void>
+
+  /**
+   * This message is used to revert the changes made by registerUsage action
```
Comment (olivermrbl) [inline packages/core/types/src/promotion/service.ts]:
```suggestion
   * This method is used to revert the changes made by registerUsage action
```

**Review by olivermrbl — COMMENTED** [review]
Campaign usage part LGTM

File: packages/modules/promotion/src/services/promotion-module.ts [hunk 3]
```diff
@@ -123,33 +123,39 @@ export default class PromotionModuleService
   @InjectManager("baseRepository_")
   async registerUsage(
     computedActions: PromotionTypes.UsageComputedActions[],
-    @MedusaContext() sharedContext: Context = {}
+    @MedusaContext()
+    sharedContext: Context = {}
```
Comment (riqwan) [inline packages/modules/promotion/src/services/promotion-module.ts]:
```suggestion
    @MedusaContext() sharedContext: Context = {}
```

File: integration-tests/modules/__tests__/cart/store/carts.spec.ts (hunk 1 above)
Comment (carlos-r-l-rodrigues) [reply integration-tests/modules/__tests__/cart/store/carts.spec.ts:2056]: I'm not sure, it seems that this is a fixed amount, but it is considering a percentage?

Comment (riqwan) [reply integration-tests/modules/__tests__/cart/store/carts.spec.ts:2056]: yeah exactly, this seems to be coming from the tax totals that are all over the place. I'll tackle this when I tackle the tax ticket for promotions.

**Comment by riqwan** [comment]
@olivermrbl @carlos-r-l-rodrigues can we get this one merged?

**Review by olivermrbl — APPROVED** [review]
LGTM, are we tackling the taxes right after this one?

**Comment by riqwan** [comment]
@olivermrbl yes, working on it right now.

**Pull request merged**